### ISSUE TITLE
Order of execution guarantee related to put & take

### ISSUE THREAD

**Issue by kheyse-oqton** [redux-saga/redux-saga#1699]
A common pattern to communicate between two sagas is to use REQUEST and SUCCESS actions.
In the caller this requires yielding a put and take. To work correctly, the take needs to be yielded before the SUCCESS action has been fired.
However due to the "asynchronous" nature of sagas, I don't know if redux-sagas actually guarantees this. Can someone give insight in this?
Can I also yield a 'select' effect between the put and take? A delay promise clearly will break it.

```javascript
    function* responderSaga() {
      yield takeEvery('SUB_PROCESS_REQUEST', function* respondSaga(action) {
        yield put({ type: 'SUB_PROCESS_SUCCESS' });
      });
    }
    function* callerSaga() {
      yield takeEvery('START', function* callSaga(action) {
        yield put({ type: 'SUB_PROCESS_REQUEST' });
        //what can be done here safely? 
        yield take(act => act.type === 'SUB_PROCESS_SUCCESS');
      });
    }
```

**Comment by Andarist** [comment]
If you have synchronous communication like this between sagas you can only perform a synchronous effects (ie `select`) before you setup take for the SUB_PROCESS_SUCCESS. 

If your success response is always synchronous you dont even have to take it though.

**Comment by kheyse-oqton** [comment]
> If you have synchronous communication like this between sagas you can only perform a synchronous effects (ie `select`) before you setup take for the SUB_PROCESS_SUCCESS.

How do I know which effects are synchronous? It does not seem to be explicitly mentioned in the docs. For some it's obvious (e.g. select), but for others not (e.g. put, put.resolve, cancel, call,...).

> If your success response is always synchronous you dont even have to take it though. // EDIT: i guess you need it if you want to receive some result from that take

I've edited the example to make clear that the respondSaga is an asynchronous saga.

**Comment by Andarist** [comment]
> How do I know which effects are synchronous? It does not seem to be explicitly mentioned in the docs. For some it's clear (e.g. select), but for others not (e.g. put, put.resolve, cancel, call,...).

What can be async? call/apply/cps (but they dont have to if called things return synchronously), `join` (if joined task is running when joining it), take (always when taking redux actions & when taking from channels when there are no buffered items on the channel), delay ofc (as its just bound call). And ofc combinators race/all when they contain mentioned effects. `fork`/`spawn` are synchronous but they might start asynchronous tasks, the effect itself resolves with task instance synchronously.

> EDIT: I have found, for instance, that yielding a put does not necessarily immediately return control back to the yielding saga, and that it may transfer to a different saga.

That's true, but a line after put gets executed synchronously in the same call stack, it just might not happen immediately - this is rather an implementation detail, unless u do really weird stuff you should be able to consider it as immediately returning effect.

**Comment by kheyse-oqton** [comment]
It would make sense to assume that both select and put are synchronous effects, but then you would expect this to not fail:
```javascript
const reducer = (state = 0, action) => (action.type === 'ACTION' ? state + 1 : state);
function* helperSaga(id) {
  const state = yield select();
  yield put({ type: 'ACTION', id });
  const newState = yield select();
  if(newState !== state + 1) throw new Error();
}
```
However, when you launch `helperSaga` multiple times like this:
```javascript
function* callerSaga() {
  yield takeEvery('START', function* callSaga(action) {
    const calls = [];
    for (let id = 0; id < 3; id += 1) {
      calls.push(call(helperSaga, id));
    }
    yield all(calls);
  });
}
```
Then the `helpersaga` will throw an error.

Maybe this is considered weird usage, but its just multiple calls, and dispatching actions conditionally based on the state is a common thing to do so I need to be sure that is the latest state.

To make it more concrete: does redux-saga guarantee this to not throw an error?
```javascript
const reducer = (state = 0, action) => (action.type === 'ACTION' ? action.id : state);
function* helperSaga() {
  yield takeEvery('ACTION', function* callSaga(action) {
    const state = yield select();
    if (state !== action.id) throw new Error();
  });
}

const N = 1000;
for (let id = 0; id < N; id += 1) {
  store.dispatch({ type: 'ACTION', id });
}
```

**Comment by shinima** [comment]
> To make it more concrete: does redux-saga guarantee this to not throw an error?
  const reducer = (state = 0, action) => (action.type === 'ACTION' ? action.id : state);
  function* helperSaga() {
  yield takeEvery('ACTION', function* callSaga(action) {
 ......


@kheyse-oqton  as far as I know, redux-saga does not guarantee this. When you dispatch an action (named A1), it may awake several sagas. These sagas continue to execute from the `yield take(...)` statement, and may dispatch several other actions (named A2, A3, A4...) synchronously. When a specific saga takes A1, all other actions (A2, A3, A4...) are already applied to the state.

Assume we have a another saga along with your helperSaga as follows...
```javascript
function* anotherSaga() {
  yield takeEvery('ACTION', function* callSaga2(action) {
    if (action.id % 2 === 1) {
      yield put({ type: 'ACTION', id: action.id * 2 })
    }
  });
}
```
... when you dispatch `{ type: 'ACTION', id: 1 }`, the anotherSaga may dispatch `{ type: 'ACTION', id: 2 }` before your helperSaga get the `{ type: 'ACTION', id: 1 }`, then the assertion that `state === action.id` would fail.

**Comment by Andarist** [comment]
> To make it more concrete: does redux-saga guarantee this to not throw an error?

Yeah, I cant think of anyway how that would broke - `dispatch` resolves `take` in `takeEvery` which in turn `fork`s `callSaga`, `select` gets executed immediately (and resolved ofc). So this should be considered 100% safe.

Your first demo thought will throw. `put`/`dispatch` gets executed in the same frame but not necessarily right away - dispatching `START` suspends the scheduler, no `put` will actually get executed until `START` gets fully processed. So you basically schedule 3 puts, but all of them executes after `START` resolves.

The whole scheduling thing is slightly tricky, but in general you shouldnt be much concerned with it.

We try to emulate concurrent programming with threads (kinda), if you consider `all` a truly concurrent effect it makes sense that state read after put might be different, because state is shared memory accessible to all threads/sagas.
